### ISSUE TITLE
[Milestones as points] refactoring

### ISSUE THREAD

**Issue by maxceem** [appirio-tech/connect-app#3299]
Follow up from comment https://github.com/appirio-tech/connect-app/issues/3210#issuecomment-531130109

From @vikasrohit :

And in fact we discussed internally to clean up the complexity involved in the milestones themselves. Here is summary of our new clean model:

1. Add new fields called `plannedDate` and `actualDate`
2. Ignore existing date fields e.g. `startDate`/`actualStartDate`/`endDate`/`completionDate`
3. Ignore `duration` field as well
4. In very simple words, a timeline would be a collection of milestones which act as simple points and does not have any duration
5. Now it is all upto the client how they want to use and render the timeline
6. We can also get rid of redundant date and duration fields on the timeline object as well

If you notice, this approach will make the logic of milestone management very simple e.g. we won't need any cascading updates when a new milestone is created or existing milestone is updated for any field. I would make the logic of progress calculation of the timeline very easy and straightforward.  It would make the movement of the milestones within timeline very simple as well.
Let me know what do you think about this. I would like to get this going and completed before we get the green signal for work management features. In our theory, backward compatibility is also not so much difficult in this case, we can simply do `plannedDate=startDate` and `actualDate=completionDate` for all existing milestones and our new rendering logic would render the timeline correctly even for the existing phases.

**Comment by vikasrohit** [comment]
I would like to start for this as soon as possible @maxceem

**Comment by maxceem** [comment]
1. Should we remove `startDate`, `actualStartDate`, `endDate`, `completionDate` and `duration` from milestones? If we keep them but don't update they would become out of date as soon as we update milestones.

2. > In very simple words, a timeline would be a collection of milestones which act as simple points and does not have any duration

   > We can also get rid of redundant date and duration fields on the timeline object as well

   If we treat milestones as points and don't have timeline end date, than we wouldn't know when the last milestone ends:
   ```
   Timeline with milestones as points:

   [community-work]----------[checkpoint-review]-------------------[final-designs]--------------------[delivery]--------(?)
   ```

    I guess we should add a new milestone type `end` which would be used to indicate the timeline end.

```
   Timeline with milestones as points (with end):

   [community-work]----------[checkpoint-review]-------------------[final-designs]--------------------[delivery]--------[end]
   ```

**Comment by vikasrohit** [comment]
1. I would like to keep them for the first release after this change so that we have a backup plan. In the next release to it, we would remove these fields if we are able to render all existing timelines correctly with new changes.

2. We don't need to know when the last milestone ends or create specific milestone type `end`. We need a new generic milestone type `reporting` which can be used for any reporting milestone like `start`, `end` or `submission ends`/`review ends` and so on. Ultimately there would always be an milestone at the end of the timeline which conveys the end of timeline.

**Comment by maxceem** [comment]
@vikasrohit on the second thoughts l'm not sure there are some benefits in refactoring milestones to `points` instead of `spans` we have now. The main issue we have with cascading updates may be solved by removing `cascading` logic from Project Service and implement an endpoint to update all the milestones for a timeline with one request so we can control the whole timeline with milestones client-side.

My main concern in moving to `points` is that we would have to control `2` objects (points) for each milestone span instead of one. Also, for each milestone span, we would need to have 2 `types` - one to indicate the **start** of milestone span and another one to indicate the **end** of milestone span. Also, it would be not so straightforward where to keep additional data for milestone points in the **start** or **end** point.

Currently, our milestone span object has the next model:
```js 
{
    id: { type: DataTypes.BIGINT, primaryKey: true, autoIncrement: true },
    name: { type: DataTypes.STRING(255), allowNull: false },
    description: DataTypes.STRING(255),
    duration: { type: DataTypes.INTEGER, allowNull: false },
    startDate: { type: DataTypes.DATE, allowNull: false },
    actualStartDate: DataTypes.DATE,
    endDate: DataTypes.DATE,
    completionDate: DataTypes.DATE,
    status: { type: DataTypes.STRING(45), allowNull: false },
    type: { type: DataTypes.STRING(45), allowNull: false },
    details: DataTypes.JSON,
    order: { type: DataTypes.INTEGER, allowNull: false },
    plannedText: { type: DataTypes.STRING(512) },
    activeText: { type: DataTypes.STRING(512) },
    completedText: { type: DataTypes.STRING(512) },
    blockedText: { type: DataTypes.STRING(512) },
    hidden: { type: DataTypes.BOOLEAN, defaultValue: false },
    deletedAt: DataTypes.DATE,
    createdAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    updatedAt: { type: DataTypes.DATE, defaultValue: DataTypes.NOW },
    deletedBy: DataTypes.BIGINT,
    createdBy: { type: DataTypes.BIGINT, allowNull: false },
    updatedBy: { type: DataTypes.BIGINT, allowNull: false },
  }
```

There are some properties like `status`, `details`, `plannedText`, `activeText`, `completedText`, `blockedText`, `hidden`, all time/user properties. They all most likely would be same for **start** and **end** point.

If we refactor to the **points** I guess most of the time we would have to update both ends (two different objects) while now we can update only one object.

Having milestones as **points** is for sure a more general approach and theoretically can cover more cases, though I don't know any cases we have or plan to have which cannot be covered using **span** milestones.

Please, let me know what you think. Is there any case, which you think we might have which cannot be covered by **spans** and can be covered by **points**, or at least it would be much easier using **points**?

**Comment by vikasrohit** [comment]
@maxceem I am not fully sure about why do we need to manage two points `start` and `end`. All, we need is to add/update a milestone as point.
I am not able to recall any use case which is not possible with *span/duration* based approach, however, as we discussed long time back and as it is written in the original description of the issue:
>we won't need any cascading updates when a new milestone is created or existing milestone is updated for any field. I would make the logic of progress calculation of the timeline very easy and straightforward. It would make the movement of the milestones within timeline very simple as well.

In nutshell:
- No cascading calculation anywhere in the system when a new milestone is added or existing one is updated (this one is huge benefit)
- Progress calculation would be `(number of milestone points in completed status)/all milestone points`

**Comment by maxceem** [comment]
@vikasrohit Yeah, that's very unlucky that we had to postpone this task for so long time. As previously this approach seams to simplify many things, though now, at a closer look I don't see such benefits anymore.

> I am not fully sure about why do we need to manage two points start and end.

In general, if we want to support milestone spans which might go in parallel (like phases in the challenges) we would have to have two points to indicate `start` and `end` (this was discussed [here](https://github.com/appirio-tech/connect-app/issues/3210#issuecomment-531169760)), example:

```js
challengeMilestones: [ // aka challenge phases
  { plannedDate: '05 January', type: 'registration-start' },
  { plannedDate: '10 January', type: 'registration-end' },
  { plannedDate: '05 January', type: 'submission-start' },
  { plannedDate: '10 January', type: 'submission-end' },
]
```

> No cascading calculation anywhere in the system when a new milestone is added or existing one is updated (this one is huge benefit)

Yeah, this was the initial intention. Though at a closer look at our real-world use-cases, it feels like we cannot avoid cascading updates even with points.

Imagine that we have a timeline with milestones as `points`. Note, that we would had to have some special `point` to indicate either `start` or `end` of the timeline, let's say we would indicate the `end`:
```js
| 05 Jan       | 10 Jan       | 15 Jan       | 20 Jan 
|--------------|--------------|--------------|
| spec         | work         | review       | end

timelineMilestones: [ // timeline of some Connect App phase/work
  { plannedDate: '05 January', type: 'phase-specification' },
  { plannedDate: '10 January', type: 'community-work' },
  { plannedDate: '15 January', type: 'checkpoint-review' },
  { plannedDate: '20 January', type: 'timeline-end' }, // as our milestones are now points, we somehow should indicate the end of last milestone span
]
```

### Add milestone to the end

Now, let's say we want to add one more milestone for final designs, which would start at the end on `20 Jan` and last for 5 days to achieve the next picture:

```js
                                             \/ added       \/ updated
| 05 Jan       | 10 Jan       | 15 Jan       | 20 Jan       | 25 Jan       
|--------------|--------------|--------------|--------------|
| spec         | work         | review       | final        | end

timelineMilestones: [ // timeline of some Connect App phase/work
  { plannedDate: '05 January', type: 'phase-specification' },
  { plannedDate: '10 January', type: 'community-work' },
  { plannedDate: '15 January', type: 'checkpoint-review' },
  { plannedDate: '20 January', type: 'final-disigns' }, // <- add new point to the end
  { plannedDate: '25 January', type: 'timeline-end' }, // <- have to update the special point which indicates the end
]
```

### Add milestone in the middle

Now, let's say we want to add one more milestone for community work before final designs. Most likely we would like to have a new milestone some duration (5 days) and all the following milestones suppose to be move forward:

```js
                                             \/ added       \/ updated     \/ updated
| 05 Jan       | 10 Jan       | 15 Jan       | 20 Jan       | 25 Jan       | 30 Jan       
|--------------|--------------|--------------|--------------|--------------|
| spec         | work         | review       | work         | final        | end

timelineMilestones: [ // timeline of some Connect App phase/work
  { plannedDate: '05 January', type: 'phase-specification' },
  { plannedDate: '10 January', type: 'community-work' },
  { plannedDate: '15 January', type: 'checkpoint-review' },
  { plannedDate: '20 January', type: 'community-work' }, // <- add new point in the middle
  { plannedDate: '25 January', type: 'final-disigns' }, // <- have to move all the following milestones
  { plannedDate: '30 January', type: 'timeline-end' }, // <- have to move all the following milestones
]
```

> Progress calculation would be (number of milestone points in completed status)/all milestone points

Currently, we are calculating timeline progress as `duration of all completed milestones / total timeline duration`, see https://github.com/appirio-tech/connect-app/blob/dev/src/helpers/projectHelper.js#L280.

> we can easily move milestone forward or backward without any need of logic to fix other milestones

If we implement some UI to drag a `point` on the `timeline` to adjust its size, than this would work. Though I guess most likely when we change the duration of some milestone span, we would want to keep other milestones spans duration untouched. So we would have to update other `points` also.

### Change duration

For example, let's say we have an initial timeline like this:
```js
| 05 Jan       | 10 Jan       | 15 Jan       | 20 Jan 
|--------------|--------------|--------------|
| spec         | work         | review       | end
```

Now, if we want to change the duration of the `work` milestone span we would still have to update `review` and `end` points:
```js
                                   \/ extended    \/ updated
| 05 Jan       | 10 Jan            | 18 Jan       | 23 Jan 
|--------------|-------------------|--------------|
| spec         | work              | review       | end
```

**Comment by vikasrohit** [comment]
Thanks @maxceem for the examples. However, I am still positive for the points based approach, here are key points:
- I guess, we don't need any demarkation of start and end of the timeline. We can assume the last milestone point in the timeline, to be the end of the timeline. If the caller wants to show a `end` milestone, it has to be added as separate milestone point. So, as soon as we remove the requirement of having a specific milestone point of type `start` or `end` , we eliminate the need of cascading for milestone movement in the timeline.
- I agree for the case where we (one front end) would treat the timeline as phases, we have to manage the cascading our selves i.e. in your last example of `Change duration` case, I would like the front end to handle this cascading. Frond end can make parallel calls to shift the `plannedDate` of all milestones next to the `work` milestone by the number of days, we want `work` phase to extend. In this case, it would make 2 parallel calls only and I don't think those calls would cause any deadlock or concurrency issue on the backend as well because they both would operate on separate resources and there is no cascading on the backend.

**Comment by maxceem** [comment]
> I guess, we don't need any demarkation of start and end of the timeline. We can assume the last milestone point in the timeline, to be the end of the timeline. If the caller wants to show a end milestone, it has to be added as separate milestone point. So, as soon as we remove the requirement of having a specific milestone point of type start or end , we eliminate the need of cascading for milestone movement in the timeline.

- I'm not sure how then do detrmine the length of all the milestone spans. Imagine that initally we want to have only 2 milestones: `phase-specification` and `community-work`. Then timelines would look like this:

   ```js
    | 05 Jan       | 10 Jan
    |--------------|
    | spec         | work
   ```

    - So `work` would have 0 `duration` in such a case. Or is there some other idea?

> I agree for the case where we (one front end) would treat the timeline as phases, we have to manage the cascading our selves i.e. in your last example of Change duration case, I would like the front end to handle this cascading. Frond end can make parallel calls to shift the plannedDate of all milestones next to the work milestone by the number of days, we want work phase to extend. In this case, it would make 2 parallel calls only and I don't think those calls would cause any deadlock or concurrency issue on the backend as well because they both would operate on separate resources and there is no cascading on the backend.

- If we keep having `milestones as spans` we also have to make 2 calls in parallel in such case. Without cascading updates on the server. So looks like there is no benefit in moving to `points` in such a case.

**Comment by vikasrohit** [comment]
> So work would have 0 duration in such a case. Or is there some other idea?

It is okay and expected to have that. I mean it is the API caller's responsibility to create correct milestones. In this particular case, API caller (connect front end) would either create a milestone called `end` OR let the user in Connect to define the correct end milestone. What I am trying to say is that, it is only matter of the view, it does not matter (theoretically) if the last milestone is called `work` or `end`, practically, some one has to create  a start and end milestone. IMO, the best candidate to do that is either the Connect front end when adding a phase, it can add default `start` and `end` milestones based on the phase start and end dates input by the user.

> If we keep having milestones as spans we also have to make 2 calls in parallel in such case. Without cascading updates on the server. So looks like there is no benefit in moving to points in such a case.

I am not suggesting to have milestones as spans but rather view milestones as spans for some particular use case like when using it in challenge timeline.

**Comment by vikasrohit** [comment]
@maxceem what do you think for this refactoring now? Am I still missing some critical possible blocker in future while thinking about the simplification?

**Comment by maxceem** [comment]
> @maxceem what do you think for this refactoring now? Am I still missing some critical possible blocker in future while thinking about the simplification?

Thank you @vikasrohit for the explanation, I think now I've got the idea more clearly. I don't see any blockers, though I also don't see immediate benefits of moving from span to points for our current cases that we have in Connect App or planning to have in works. Anyway, in general, points would be a more universal way of storing milestones.

There are 2 main moments to confirm.

## How to migrate existent **span** milestones

As in our UI we are showing only starting dates for milestones, so we can migrate each of milestone "span" to a new milestone "point" keeping the only starting dates. And additionally, add one milestone with type `end` to each timeline to indicate the end of the last milestone:

![Milestones refactoring (2)](https://user-images.githubusercontent.com/146016/74003797-d383f880-49ae-11ea-963c-c6e3e3041aac.png)

So for migrating, we would have to:
- For each existent milestone:
   - copy `startDate` to `plannedDate`
   - copy `actualStartDate ` to `actualDate`
- For each existent timeline with at least one milestone:
   - add one more milestone with:
      - `type` as `"end"`
      - `plannedDate` as `endDate` of the last milestone in timeline
      - `actualDate` as `completionDate ` of the last milestone in timeline

## Bulk milestone operations

To avoid cascading updates we would have to implement endpoint to update multiple milestones at a time. We cannot avoid having such endpoint neither with "spans" nor with "points".

Endpoint `PATCH /v5/timelines/:timelineId(\\d+)/milestones` should accept a list of milestones to create and/or update.

This endpoint can support updates of existing milestones, and creating new milestones. Example body request:
```js
[
   {
      // no id this milestone would be created
      "name": "New milestone",
      "plannedDate": "12 Dec 2020",
   },
  {
      "id": 13, // has id, this milestone would be updated
      "name": "Existent milestone 1",
      "plannedDate": "15 Dec 2020",
   },
  {
      "id": 14, // has id, this milestone would be updated
      "name": "Existent milestone 2",
      "plannedDate": "20 Dec 2020",
   }
]
```
Example response may have HTTP status `200` or `207 - Multy Status`, see https://tools.ietf.org/html/rfc4918#section-11.1. While per each record we would additionally return status:
```js
[
   { // succesfully created milestone
      "status": 200,
      "data": {
         "id": 12,
         "name": "New milestone",
         "plannedDate": "12 Dec 2020",
      }
   },
  { // succesfully updated milestone
     "status": 200,
     "data": { 
         "id": 13, // has id, this milestone would be updated
         "name": "Existent milestone 1",
         "plannedDate": "15 Dec 2020",
      }
   },
  { // failed to update milestone
      "status": 404,
      "error": {
        "message": "Milestone with id 14 is not found.",
        "debug": "...."
      }
   }
]
```

@vikasrohit what do you think, would this work for us?

**Comment by vikasrohit** [comment]
> I also don't see immediate benefits of moving from span to points for our current cases

I would like to proceed further, only when we all agree on the fact that points based implementation makes our life easy especially w.r.t. cascading of milestone changes.

For backward compatibility, I think my original statement still holds true, let me know if there is any problem with that:
> In our theory, backward compatibility is also not so much difficult in this case, we can simply do plannedDate=startDate and actualDate=completionDate for all existing milestones and our new rendering logic would render the timeline correctly even for the existing phases.

> type as "end"

I don't think we need a specific type of milestone to mark the end of the timeline. In your specific example, `Delivery` would automatically be the end of the timeline as it is the last visible milestone in the timeline.

**Comment by maxceem** [comment]
> I would like to proceed further, only when we all agree on the fact that points based implementation makes our life easy especially w.r.t. cascading of milestone changes.

My main concern is that for avoiding cascading updates we have to intoroduce a **bulk** endpoint for updating multiple milestones at a time. And we would need it for any of the approaches with "points" or with "spans".

> I don't think we need a specific type of milestone to mark the end of the timeline. In your specific example, Delivery would automatically be the end of the timeline as it is the last visible milestone in the timeline.

So far each of the milestone "spans" have its duration including delivery. We use it to show the progress or delay of each milestone:

![image](https://user-images.githubusercontent.com/146016/74030953-46ab6000-49eb-11ea-944f-6f10a7ce9d2b.png)

If **delivery** is the last **point** then we cannot show the progress/delay for it.

Also, we use milestone "spans" duration to calculate the total timeline/phase duration. If we don't add additional "end" milestone, then timeline duration would be calculated not the same way as before. The timeline duration would be reduced by the duration of last milestone, as it would not be taken into account:

![Milestones durations](https://user-images.githubusercontent.com/146016/74031522-ac4c1c00-49ec-11ea-88cd-a9794fdaffc9.png)

**Comment by vikasrohit** [comment]
> My main concern is that for avoiding cascading updates we have to intoroduce a bulk endpoint for updating multiple milestones at a time. And we would need it for any of the approaches with "points" or with "spans".

I was thinking of making parallel requests, for the first release, instead of batching them because most of the times, we would have about 4-5 milestones to be updated after one of the early milestone is updated. However, after reading more about having bulk endpoints in REST api, I think we can have one for this case. I think that is good to go.

For showing progress or delay of the last milestone (or infact for any milestone), I guess we can now count the progress based on the milestone's planned date and current date e.g. in your example, delivery milestone would show the delay of 6 days if the current day is 6 days after the Delivery milestone's planned date (which we have copied as `startDate` from the existing span based milestone). 

For the difference in timeline duration, I think first we need to update our backward compatibility script to as follows:

```
plannedDate       = startDate + duration
actualDate          = completionDate
```

And I guess, we don't need to add the `end` milestone, but the `start` milestone because our new point based milestones are completion of the task. We can add a new milestone called `Start` (note, it is not different than other milestone other than text) and rename the existing `Start` milestones to read something like `Planning`/`Requirement Analysis` (I like the `Planning`).
So, an old timeline like:
```
**Name      startDate  completionDate**
Start        01-02        03-02
Work1     03-02        07-02
Work2    07-02        09-02
Delivery 09-02        10-02
```
Would transform into
```
**Name      plannedDate  actualDate**
Start           01-02               01-02
Planning    03-02               03-02
Work1        07-02               07-02
Work2       09-02                09-02
Delivery    10-02                10-02
```

**Comment by maxceem** [comment]
Thank you for detailed reply @vikasrohit.

`Start` milestone would also work. Though it would a little bit harder to adapt the existent UI to use "poins" with `Start` milestone instead of an `End` one.

The reason is that in our UI we are showing the `startDate` as dates, but in "points" we would hold the date when a milestone is finished, not started:

![image](https://user-images.githubusercontent.com/146016/74224384-b3787000-4cf3-11ea-8946-5d614978dbf4.png)

I think we can still do that, and use `Start` milestone instead of `End`. Though I think we shouldn't show this `Start` milestone in our current timeline. Why do I think we shouldn't show it?
First of all to keep the current UI as it is.
The second thing, the `Start` milestone would be always started and completed at the same time in our workflow.

Imagine we have a timeline which is intially created with phase:

```
**Name      plannedDate     actualDate        status**
Start       01-02               01-02          Planned
Planning    03-02               03-02          Planned
Work1       07-02               07-02          Planned
Work2       09-02               09-02          Planned
Delivery    10-02               10-02          Planned
```

As soon as we move a phase to `In Progress` status we have to mark `Start` milestone as completed and activate `Planning` milestone, so it would like this:

```
**Name      plannedDate     actualDate        status**
Start       01-02               01-02          Completed <<< as timeline has started
Planning    03-02               03-02          Activate <<< as this is the current milestone
Work1       07-02               07-02          Planned
Work2       09-02               09-02          Planned
Delivery    10-02               10-02          Planned
```

**Comment by vikasrohit** [comment]
@maxceem I don't think we can keep the UI as is with new points base implementation because to do that we have to store the some kind of `start` date for each milestone which we don't have right now. As you said, we can show the `completionDate` of the previous milestone to render it in this way, however, I think that would be confusing technically and would stop us from fully utilizing the power of simplicity of new model for new timelines. So, I guess, we should be okay as long the new rendering of the timeline depicts the correct time of events. That is why I would like to have this development in separate branches so that we can decide when to merge these into main branch. I would discuss about these changes with product managers when we actually have it working in our parallel env/branches. We won't do the migration of the whole data even in dev until we are sure that it would work for us by migrating a sample of timelines to the new model (for the time being we can create new table instead of modifying the same table).

> As soon as we move a phase to In Progress status we have to mark Start milestone as completed and activate Planning milestone, so it would like this:

That is good, I guess. I mean logically it means phase has **started**. The only thing is that we might not need to `Activate` the `Planning` milestone as we can by default assuming the first non hidden and non completed milestone to be active in UI. It seems redundant to me now to have status like `Active`, however, I am not sure how it would work with our another feature of `pausing` a milestone work.

**Comment by maxceem** [comment]
> I don't think we can keep the UI as is with new points base implementation because to do that we have to store the some kind of start date for each milestone which we don't have right now.

If we don't show the start date of milestones how the timeline should look when using points?

![image](https://user-images.githubusercontent.com/146016/74313232-44138680-4dae-11ea-8c1b-a73acf2a4dff.png)

@vikasrohit I'd like to clarify one moment. We are going to move from `phases` with timelines to `works` with timelines, while keeping `phases` with timelines for old projects. The question is, as soon as we enable `workstreams` and new projects would be always created with workstreams, do we plan to make any improvements on how timelines work in `phases`? Or as soon as we enable `works` we would only continue and improve how timelines work inside `works`, but we would keep timelines in `phases` with the same functionality as we have now without improvements?

**Comment by vikasrohit** [comment]
> If we don't show the start date of milestones how the timeline should look when using points?

I am assuming the dates in your example timeline are `startDate` field of the current milestones:
```
Start       10-01               10-01          Planned
Planning    12-01               12-01          Planned
Work       26-01               26-01          Planned
Delivery    27-01               27-01          Planned
```

For the `phases` vs `works` question, I don't see much changes, except bug fixes, for the phase's timelines once we have `works` in place. However, I don't understand how that affects our current work or future efforts because I think once we have the milestones a points, all the changes that might be needed in `phases` would be limited to front end not to the backend.

**Comment by maxceem** [comment]
Thank @vikasrohit.

Let's try to visualize the UI of such a timeline as per the data above. Imagine that we changed the status of the Phase to `In Progress`. So the  **Start** milestone point becomes `completed` and **Planning** milestone point becomes `active`. How this should look in our UI?

#### Version 1

![image](https://user-images.githubusercontent.com/146016/74405866-22c3a080-4e69-11ea-81f3-daaccb823134.png)

- This picture as per current implementation. It represents the timeline **incorrectly** as it's showing as **Planning** milestone has started on `12 Jan` though **Planning** milestone should finish on `12 Jan` and start on `10 Jan`.

#### Version 2

So it looks like we have to change the UI to show the progress **before** the active milestone, as active milestone shows the `endDate`, not the `startDate`:

![image](https://user-images.githubusercontent.com/146016/74406417-91edc480-4e6a-11ea-97c3-b5398af6cd4f.png)

- In such a case, dates look good. We can see that we at a milestone now that started on `10 Jan` and should finish on `12 Jan` which is good.
- Though the rest of the things looks **incorrect**. From this UI it looks like the current milestone is **Start**, not **Planning**.

Should we do more changes to UI somehow?

**Comment by vikasrohit** [comment]
<img width="597" alt="Screen Shot 2020-02-13 at 5 29 06 PM" src="https://user-images.githubusercontent.com/2417632/74433404-58c34d80-4e86-11ea-9a86-942a605b2df1.png">
@maxceem do you think something like this would help?

**Comment by maxceem** [comment]
> @maxceem do you think something like this would help?

Yes, we can do like this, we only have to decide whether we show the date in front of the small blue circle (position 1) or in front of the last line of the milestone (position 2), as some milestones look like this:

![image](https://user-images.githubusercontent.com/146016/74617490-41c37a80-5168-11ea-89d6-788f00e3ad49.png)

Though I'm not sure, what is the advantage of using milestone points with `end` dates in favor of milestones points with `start` dates? The benefit of switching to "point" based model is that the client may decide on its own how to use the data of milestone points. It looks like milestone points with `start` dates fit the current UI pretty close. So we wouldn't need to change the existing UI and changes required on the client would be minimal.

**Comment by vikasrohit** [comment]
I think both are fine for me, personally I like the position 1 as that shows the latest action that has to be performed to reach the milestone but I don't have any negative for the position 2 as well. I can discuss this with product managers when we have other technical details sorted out.

> Though I'm not sure, what is the advantage of using milestone points with end dates in favor of milestones points with start dates? The benefit of switching to "point" based model is that the client may decide on its own how to use the data of milestone points. It looks like milestone points with start dates fit the current UI pretty close. So we wouldn't need to change the existing UI and changes required on the client would be minimal.

Looks like we are coming to same page now. :)

Let me know what else tricky situations or question we need to handle before we start the implementation, I need to present this change to the product managers before we start working on this.

**Comment by maxceem** [comment]
> Let me know what else tricky situations or question we need to handle before we start the implementation, I need to present this change to the product managers before we start working on this.

To sum up, there are two main questions which we were discussing.

### If we are migrating to "points", how we migrate existent timelines?

We've discussed two main approaches:
1. Treat milestone points as "start" dates and add an additional `End` milestone as was proposed here https://github.com/appirio-tech/connect-app/issues/3299#issuecomment-583248902

2. Treat milestone points as "end" dates and add an additional `Start` milestone as proposed here https://github.com/appirio-tech/connect-app/issues/3299#issuecomment-584564474

Personally, I'm in favor of the 1st approach, as in such case we can keep UI as it is without any changes. And the changes which we have to do client-side for migrating would be minimal in such a case.

I don't see what advantages we would get in the second approach in comparison with using "start" dates as we would have to modify the current UI which may lead to more decisions to make and it would require more changes client-side to migrate to this way. Though this approach should also work and it's also quite clear for me.

### What are the benefits in migrating to "points"?

In general, while I think migrating to points should be quite smooth, I still don't see strong immediate advantages in migrating to points milestones for our cases. Milestones as spans seem to also fit our current needs. To achieve our main aim to **get rid of cascading updates** we would use Bulk Update API endpoint no matter if use "points" or "spans".
Though potentially, having milestones as points we would be more free in implementing any logic.

Maybe if we continue working on timelines for "works" there would appear clear situations that would benefit from "point" milestones, though it's hard for me to see it at this point of time. Maybe we may postpone it for a while and get rid of cascading updates first while still keeping "span" milestones, and after we would see if there are any other issues we come across which could be resolved by migrating to "point" milestones.

**Comment by vikasrohit** [comment]
> I don't see what advantages we would get in the second approach in comparison with using "start" dates as we would have to modify the current UI which may lead to more decisions to make and it would require more changes client-side to migrate to this way. Though this approach should also work and it's also quite clear for me.

Oh, I interpreted your statement previous to this statement that you are now in favour of approach 2. :)
The major reason to go for approach#2 is that it is logical in terms of milestone's meaning. Milestone means something has achieved at that point. So, I would like to go for milestones points as end dates.

> Maybe we may postpone it for a while and get rid of cascading updates first while still keeping "span" milestones, and after we would see if there are any other issues we come across which could be resolved by migrating to "point" milestones.
I think this makes sense for now as we want to get rid of non v5 standards as soon as possible. And as we discussed that we need bulk endpoint for both approaches, I am now inclined towards getting rid of cascading updates first with existing approach. Please feel free to create separate issue (if does not exists) and start working on that. We will take care of this issue later.
